src/agent/onefuzz/src/syncdir.rs

```diff
@@ -0,0 +1,134 @@
+        match fs::metadata(&self.path).await {
+            Ok(m) => {
+                if m.is_dir() {
+                    Ok(())
+                } else {
+                    anyhow::bail!("File with name '{}' already exists", self.path.display());
+                }
+            }
+            Err(_) => fs::create_dir(&self.path).await.map_err(|e| e.into()),
+        }
+    }
+
+    pub async fn continuous_sync(
+        &self,
+        operation: SyncOperation,
+        delay: Option<Duration>,
+    ) -> Result<()> {
+        let delay = delay.unwrap_or(DEFAULT_CONTINUOUS_SYNC_DELAY);
+        loop {
+            self.sync(operation).await?;
+            tokio::time::delay_for(jitter(delay)).await;
+        }
+    }
+
+    async fn file_uploader_monitor(&self, event: Event) -> Result<()> {
+        let url = self.url.url();
+        verbose!("monitoring {}", self.path.display());
+
+        let mut monitor = DirectoryMonitor::new(self.path.clone());
+        monitor.start()?;
+        let mut uploader = BlobUploader::new(url);
+
+        while let Some(item) = monitor.next().await {
+            event!(event.clone(); EventData::Path = item.display().to_string());
+            if let Err(err) = uploader.upload(item.clone()).await {
+                bail!(
+                    "Couldn't upload file.  path:{} dir:{} err:{}",
+                    item.display(),
+                    self.path.display(),
+                    err
+                );
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Monitor a directory for results.
+    ///
+    /// This function does not require the directory to exist before it is called.
+    /// If the directory is reset (unlinked and recreated), this function will stop
+    /// listening to the original filesystem node, and begin watching the new one
+    /// once it has been created.
+    ///
+    /// The intent of this is to support use cases where we usually want a directory
+    /// to be initialized, but a user-supplied binary, (such as AFL) logically owns
+    /// a directory, and may reset it.
+    pub async fn monitor_results(&self, event: Event) -> Result<()> {
+        loop {
+            verbose!("waiting to monitor {}", self.path.display());
+
+            while fs::metadata(&self.path).await.is_err() {
+                verbose!("dir {} not ready to monitor, delaying", self.path.display());
+                tokio::time::delay_for(jitter(DELAY)).await;
+            }
+
+            verbose!("starting monitor for {}", self.path.display());
+            self.file_uploader_monitor(event.clone()).await?;
+        }
+    }
+}
+
+pub async fn continuous_sync(
+    dirs: &[SyncedDir],
+    operation: SyncOperation,
+    delay: Option<Duration>,
+) -> Result<()> {
+    let delay = delay.unwrap_or(DEFAULT_CONTINUOUS_SYNC_DELAY);
+    loop {
+        for dir in dirs {
+            dir.sync(operation).await?;
+        }
+        tokio::time::delay_for(jitter(delay)).await;
```

Comment:
Maybe we should define and export a `delay_with_jitter()` helper function from the `crate::jitter` module.